src/main/java/frc/robot/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,93 @@
+import com.revrobotics.SparkPIDController;
+
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class TurretSubsystem extends SubsystemBase {
+
+  private CANSparkMax turretMotor;
+  private SparkPIDController turretPIDController;
+  private RelativeEncoder m_encoder;
+
+  private double robotAngle = 90;
+  private double robotXOffset = 10;
+private double robotYOffset = 10;
+
+  public TurretSubsystem() {
+    turretMotor = new CANSparkMax(Constants.TurretConstants.TURRET_MOTOR_ID, MotorType.kBrushless);
+    turretMotor.restoreFactoryDefaults();
+    turretMotor.setSmartCurrentLimit(40);
+    
+    turretMotor.restoreFactoryDefaults();
+
+    turretMotor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kForward, true);
+    turretMotor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, true);
+
+    turretMotor.setSoftLimit(CANSparkMax.SoftLimitDirection.kForward, 1);
+    turretMotor.setSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, -1);
+
+    // initialze PID controller and encoder objects
+    turretPIDController = turretMotor.getPIDController();
+    m_encoder = turretMotor.getEncoder();
+
+    // set PID coefficients
+    turretPIDController.setP(Constants.TurretConstants.TURRET_KP);
+    turretPIDController.setI(Constants.TurretConstants.TURRET_KI);
+    turretPIDController.setD(Constants.TurretConstants.TURRET_KD);
+    turretPIDController.setIZone(Constants.TurretConstants.TURRET_KIZ);
+    turretPIDController.setFF(Constants.TurretConstants.TURRET_KFF);
+    turretPIDController.setOutputRange(Constants.TurretConstants.TURRET_KMINOUTPUT, Constants.TurretConstants.TURRET_KMAXOUTPUT);
+
+    turretPIDController.setSmartMotionMaxVelocity(Constants.TurretConstants.TURRET_MAXVEL, Constants.TurretConstants.TURRET_SMARTMOTION_SLOT);
+    turretPIDController.setSmartMotionMinOutputVelocity(Constants.TurretConstants.TURRET_MINVEL, Constants.TurretConstants.TURRET_SMARTMOTION_SLOT);
+    turretPIDController.setSmartMotionMaxAccel(Constants.TurretConstants.TURRET_MAXACC, Constants.TurretConstants.TURRET_SMARTMOTION_SLOT);
+    turretPIDController.setSmartMotionAllowedClosedLoopError(Constants.TurretConstants.TURRET_ALLOWEDERR, Constants.TurretConstants.TURRET_SMARTMOTION_SLOT);
+
+  }
+
+  public void setTurretPos(double pos) {
+    turretPIDController.setReference(pos, CANSparkMax.ControlType.kSmartMotion);
+  }
+
+  public void setTurretVelo(double velo){
+    turretPIDController.setReference(velo, CANSparkMax.ControlType.kVelocity);
+  }
+
+  public double getTurretPos(){
+    return turretMotor.getEncoder().getPosition();
+  }
+
+  @Override
+  public void periodic() {
+    
+  }
+
+  public double turretFromOdometry(Pose2d robotPos){
+    return Math.atan(robotPos.getY()/robotPos.getX()) + robotPos.getRotation().getDegrees();
```

Comment:
Not sure what the intention of this is?  Can you please describe it further

Reply:
You will have to ask rishik because I am lost as to what this does.

Reply:
It calculates how much the turret has to move given the position of the robot by using trig to calculate the X and Y offset from the target and then the offset angle of the robot and add that together to calculate the amount the turret has to move